willow/plugins/pillow.py

```diff
@@ -246,4 +254,144 @@ def to_buffer_rgba(self):
+
+    @Image.operation
+    def rotate(self, angle):
+        return PillowAnimatedImage([frame.rotate(angle) for frame in self.frames])
+
+    @Image.operation
+    def set_background_color_rgb(self, color):
+        return PillowAnimatedImage([frame.set_background_color_rgb(color) for frame in self.frames])
+
+    @Image.operation
+    def save_as_jpeg(self, f, quality=85, optimize=False, progressive=False):
+        if self.has_animation():
+            pass  # TODO: Raise warning
+
+        return self.frames[0].save_as_jpeg(f, quality=quality, optimize=optimize, progressive=progressive)
+
+    @Image.operation
+    def save_as_png(self, f, optimize=False):
+        if self.has_animation():
+            pass  # TODO: Raise warning
+
+        return self.frames[0].save_as_png(f, optimize=optimize)
+
+    @Image.operation
+    def save_as_gif(self, f):
+        image = self.frames[0].image
+        frames = self.frames
+
+        # All gif files use either the L or P mode but we sometimes convert them
+        # to RGB/RGBA to improve the quality of resizing. We must make sure that
+        # they are converted back before saving.
+        if image.mode not in ['L', 'P']:
+            frames = [
+                frame.convert('P', palette=_PIL_Image().ADAPTIVE)
+                for frame in frames
+            ]
+
+        params = {
+            'save_all': True,
+            'duration': image.info['duration'],
+            'append_images': [frame.image for frame in frames[1:]]
+        }
+
+        if 'transparency' in image.info:
+            params['transparency'] = image.info['transparency']
+
+        image.save(f, 'GIF', **params)
+
+        return GIFImageFile(f)
+
+    @Image.operation
+    def save_as_webp(self, f):
+        if self.has_animation():
+            pass  # TODO: Raise warning
+
+        return self.frames[0].save_as_png(f, optimize=optimize)
+
+    @Image.operation
+    def auto_orient(self):
+        # Animated GIFs don't have EXIF data
+        return self
+
+    @classmethod
+    @Image.converter_from(GIFImageFile)
+    def open(cls, image_file):
+        image_file.f.seek(0)
+        image = _PIL_Image().open(image_file.f)
+
+        frame = image
+        frames = []
+        while frame:
+            frames.append(frame.copy())
+
+            try:
+                foo = image.seek(image.tell() + 1)
+            except EOFError:
+                break
+
+        return cls([PillowImage(frame) for frame in frames])
+
+    @Image.converter_to(RGBImageBuffer, cost=200)
+    def to_buffer_rgb(self):
+        if self.has_animation():
+            pass  # TODO: Raise warning
+
+        image = self.image
+
+        if image.mode != 'RGB':
+            image = image.convert('RGB')
+
+        return RGBImageBuffer(image.size, image.tobytes())
+
+    @Image.converter_to(RGBAImageBuffer, cost=200)
+    def to_buffer_rgba(self):
+        if self.has_animation():
+            pass  # TODO: Raise warning
```

Comment:
Just to clarify: are you planning to fix these todos separately or in this PR?

Reply:
Ahh, forgot I had these. I'll have another look in this PR

Reply:
Do we care if it is an animated gif? It is common practice to get a static image when saving to png/jpeg

Reply:
If Willow is forced to decide which frame to use when saving the image, I think this should raise a warning. I know it's common to just pick the first frame but I think it's better to require them to explicitly state this as they might be calling `.save_as_png` without considering what to do with animated GIFs.

It also never makes sense to attempt to do face detection on an animated image. So if this happens there must be a missing check and a warning should help surface this issue.

We probably should have an operation available on all image classes to allow selecting individual frames in the image. This way, users can always explicitly choose the first frame if they want to remove animation (and this would have no effect on non-animated images).

Something like:

```python
myimage.frame(0).save_as_png()
```

Reply:
Choosing a frame sounds like a good idea. Should probably be a separate PR to keep this focussed.